### ISSUE TITLE
Getting virtual gamepads to play nice inside the runtime

### ISSUE THREAD

**Issue by HugLifeTiZ** [ValveSoftware/steam-runtime#422]
#### Your system information

* Steam Runtime Version: v0.20210518.0-18-g43b6f7b
* Distribution (e.g. Ubuntu 18.04): Xubuntu 20.04
* Link to your full system information (Help -> System Information) in a [Gist](https://gist.github.com/): https://gist.github.com/HugLifeTiZ/69744af85762b661b68214645607fc4b
* Have you checked for system updates?: Yes
* What compatibility tool are you using?: Proton 5.13+
* If you are using Proton 5.13 or newer: What versions are listed in `SteamLinuxRuntime_soldier/VERSIONS.txt`? v0.20210518.0-18-g43b6f7b, pressure-vessel 0.20210518.0+srt1, soldier 0.20210505.0

#### Please describe your issue in as much detail as possible:

Hi there. I know my tweaked setup has caused myself and Flatpak/Steam maintainers a few headaches. I am currently going through one right now, because in order to play Guilty Gear Strive, at present, I have to rip out my entire input abstraction setup that has been working well for me for years.

So I'll give some background on myself, what I'm doing, and why I'm doing it. I have been involved in the fighting game community since 2012. I have been a Linux user since 2008, but back when I got involved with FGC, gaming on my Linux computer was simply not happening. Until TowerFall came out in 2014. For a couple years, I was bringing four controllers with me anywhere someone might want to play TowerFall. Parties, FGC events, etc. But the FGC people wanted to use their own controllers, and that complicated things.

In 2014, we were still mostly in the era of where the only controller that mattered was the Xbox 360 controller. And controllers connecting to the computer could be anything. A first-party 360 pad, a dualshock 3, a third party Xinput recognized or not recognized by xpad, a third-party PS3 controller, a controller that is just generic HID, etc etc etc. So like everyone else at the time (and unfortunately, still in 2021 too), I used xboxdrv to spoof 360 controllers, along with [a bash script](https://github.com/HugLifeTiZ/rejoy) to automate it and generate mappings, until [MoltenGamepad](https://github.com/jgeumlek/MoltenGamepad) came out.

MoltenGamepad's main draws were that you could provide generic mappings that provided dumb "heuristics" for figuring out what a controller is and how to translate it to the 360 layout, you could have the virtual pads always present to "fix" hotplugging and prevent device gaps on games that react stupidly to them, and you could map to keyboard and mouse at the event device level. Anyways, after contributing a number of generic device mappings, every controller my laptop has ever seen has simply worked.

But then you have Steam Input that is doing the same thing but within a walled, closed-source ecosystem that I couldn't configure as much, so I had to sort of neuter it by making sure that the user that MG ran as, a system user I called gamepad, was the only one that could see event and hid devices for real gamepads. Steam Input creating an extra abstraction over four virtual 360 pads is fine. Except that it seems that in certain circumstances, it stops doing that. That's why I'm here today, but I'll get back to that because I feel compelled to finish the background exposition.

A few years ago, Steam Play and Proton hit the scene, with Tekken 7 on the whitelist. Tekken 7! I could hardly believe it! My main game in 2012 was SoulCalibur V and I was very badly jonesing for a 3D FG experience, so I was very much in on Tekken 7 on my PS4. But I could play it on my PC? Tekken 7?! Well, to tell the truth, I had been trying it out as far back as when DXVK hit the scene, and it was *okay*, but Tekken 7 being on the whitelist elevated the exprience incredibly, and blew the doors wide open for integrating my laptop into the scene. And because I had MoltenGamepad handling my controllers, anyone could plug in their fight stick, their Hori fighting commander, their Pokken pad, their PS2 controller with a Brook adapter, and it would all just work. And then when SoulCalibur VI came out, after a bit of trepidation, I started playing it on my laptop as well. Along with OBS recording our sessions, [it has been an integral part of our scene since.](https://www.youtube.com/playlist?list=PL_iI2KUM3Q-rZ3MZwDqytQ9MmJTFEX4ti)

I have built scripts around MoltenGamepad to monitor when it finds devices and loses devices, to manage mapping and translation profiles, and to swap slots when necessary. Most recently, I made a script for [pairing bluetooth controllers similarly to how a console does](https://github.com/HugLifeTiZ/configs-scripts/blob/master/scripts/pair-controllers), and a [nice little menu](https://github.com/HugLifeTiZ/configs-scripts/blob/master/scripts/game-menu) for people to be able to disconnect their DS4s and let someone else get in on the slot they were just in. It doesn't seem that Steam BPM has any means to pair new bluetooth controllers, and this experience is better by far than opening Blueman to pair new controllers. Especially since I'm the only one who can be expected to know how to do that. In a fighting game scene, players and controllers come and go on a given device relatively frequently. Right now (or rather, last time I checked), pairing by connecting over USB doesn't work if the controller paired with another device but is still remembered by the computer. It also brings up a notification prompt unless you have an agent running that just accepts all requests. So at present, dealing with a large volume of controllers that come and go is a suboptimal experience by default, improved by my scripts.

Right, so that's a lot of exposition for my background, but I think that's enough, so now let me try to get to the point of why I'm here trying to start a discussion. I am very bad at getting to the point, so I apologize.

Guilty Gear Strive is out, so I was hoping it would work, but I hit a stumbling block at the title screen. My controller wasn't working. I posted about it, but it seemed others weren't having that issue. After some testing of different Proton versions, I discovered that the last Proton-GE before it started building against Soldier, 5.9-GE-ST, could see my controller, but it froze at the character select screen. I remembered that another game had a similar situation, Bloodstained: Ritual of the Night. It also can't use my controller on any Proton built against Soldier.

Let me be more technically correct. *It can't use MoltenGamepad's virtual devices.* If I disable MG and the udev rules that hide real devices (actually, I patched them to do nothing if MG isn't running, but I never submitted those changes upstream), then it can see my DualShock 4 and use it just fine. But that didn't make sense to me. Tekken 7 works. SoulCalibur VI works. Power Rangers works *if I disable Steam Input entirely.*

I'm just an end user and relatively inexperienced developer, but I think I might be seeing what's going on. Steam Input is not working with virtual devices inside Soldier at all, for some reason. But XInput, on the other hand, is working just fine with virtual devices. Yet inside Steam itself, if I pull up the controller configuration menu, it sees the four virtual devices MG makes. It enumerates them backwards for some reason (actually, a lot of Flatpak apps do), but they do work. So that suggests to me that with the combination of Steam Input and the Soldier runtime, virtual devices are being ignored. If the game can fall back onto XInput or some other means of reading a controller, everything is fine. But if a game uses Steam Input and *only* Steam Input, it won't work, because the virtual controllers are not being handled by it. Is there a particular reason for that? Is there anything I can do to get around it? Am I understanding this situation correctly at all? It does not seem that you are the only ones ignoring virtual devices; [it's a problem with Chromium too](https://github.com/jgeumlek/MoltenGamepad/issues/93) that I ended up [reporting upstream](https://bugs.chromium.org/p/chromium/issues/detail?id=1171769), but they haven't updated on that in a bit.

I understand that you may not be inclined to support virtual devices since Steam Input generates virtual devices itself. MoltenGampead is largely redundant with Steam Input, but it runs outside of Steam, and I can do much more with it. I've built a good experience for my local fighting game scene with it. If I were to simply stop running MoltenGamepad, what would I lose? Well, that menu I built comes up when I press the Guide button. My default profile maps that to Super+G, which my DE uses to run the command for the menu. Then I ask MG to maps gamepad inputs to keyboard so rofi can be controlled via gamepad. In order to get back that functionality, I'd need to basically create a brand new scriptable menu that is natively controllable via gamepad, or add AntiMicroX (or something like it) into the workflow. And I'd need to intercept the guide button somehow, find a way to make sure applications ignore it, or stop using it. Intercepting it can only be done via event devices, and would result in a virtual device that is, again, ignored. I know with DualShock controllers, you're watching the hidraw device; I don't think we can do anything with that.

What about other usecases for virtual devices that *don't* involve replicating your input abstraction infrastructure? Well, let's ask this: do you all support joycons yet? MoltenGamepad does support joycons, both in single and paired modes. But even without MoltenGamepad, there's [another daemon](https://github.com/DanielOgorchock/joycond) that handles these complex devices, but in order to combine them, the result is, you guessed it: a virtual device.

I strongly believe that virtual devices should be supported. If they're being explicitly ignored, I would like to request that you change that. If they're not being ignored on purpose, then I guess it's troubleshooting time, and I would like to request your assistance in that endeavor. Are there undocumented environment variables that would help me here?

I know this post was a lot, so if you read it, thanks for your time and effort. I want to make sure you know how much I appreciate all of you and everything you've done for the Linux gaming ecosystem.

**Comment by smcv** [comment]
Nothing in the Steam Runtime or pressure-vessel is *deliberately* ignoring virtual gamepads, but there are a lot of layers here: the Steam client, SDL2 and/or Proton's version of Wine might be deliberately ignoring virtual gamepads, or something might be *accidentally* ignoring them.

I left some background information on <https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/issues/72#note_20862> about how the various components work.

Looking in `/proc/*/fd` would tell you whether the various programs involved have the relevant device node open. Launching games with `PRESSURE_VESSEL_SHELL=instead` in the environment (which runs an `xterm` instead of the actual game, with the actual game runnable as `"$@"`) might be helpful for debugging; the `steam-runtime-input-monitor` utility available in the container does more or less the same enumeration that SDL and Proton's Wine do (although it's a parallel implementation and might differ in some details).

Because you said "XInput" works and Steam Input does not, this sounds like more of a `steam-for-linux` issue than a `steam-runtime` issue, although the pressure-vessel container does change things by making it impossible to enumerate devices by talking to udevd (the udevd <-> libudev protocol isn't container-friendly), which results in SDL and Proton's Wine having to fall back to a simpler device enumeration mechanism involving reading `/dev` and `/sys` themselves.

The developers of the Steam Runtime cannot tell you how Steam Input works: that's part of the proprietary Steam client and the Steamworks API.

"XInput" presumably means Proton's Wine implementing low-level device discovery in `dlls/winebus.sys`, and then trying to expose the resulting devices in the same way that XInput would on Windows.

> It enumerates them backwards for some reason (actually, a lot of Flatpak apps do)

Device enumeration order is arbitrary/undefined. If two components want to enumerate pre-existing controllers in the same order, then they would need some way to cache a list of devices during initial "coldplug" enumeration, sort the list in some order (I suppose they could maybe use the mtime of the device nodes), then behave as though the devices had been hotplugged in the desired order. At the moment, neither SDL nor Proton/Wine makes any attempt to do that.

**Comment by HugLifeTiZ** [comment]
Thank you very much for your detailed and informative replies to both this issue and my other issue on the other tracker. Again, my apologies for polluting the other tracker.

I have done a lot of experimentation, and I have found a few things. One important thing I found was that I had separate, different problems conflated with each other, and masking each other a little bit. The problems I am having with Bloodstained: ROTN and Hades are actually different from the problems I was having with Guilty Gear Strive.

Problem 1: unfortunately, device enumeration order is significant, and it's breaking some games. In sandboxed environments, controllers are enumerated in reverse from how they would be enumerated outside of the sandbox.

I primarily used Bloodstained: ROTN and Hades for testing, because I found that Hades also worked in newer Proton versions if I turned off MoltenGamepad. I previously could only play Hades using an older Proton and the distro Steam package, and a buttload of launch options. Guilty Gear Strive is what set me out on this journey, but starting it up is just... so damn annoying. I don't want to spend 2 minutes "communicating with server" before the title screen every time I want to see if my controllers are seen, so I used Strive for only the last few tests.

One important thing that I noticed is that whenever I have MoltenGamepad off, the lack of abstraction means that there is only one controller ever visible to the system. There is no ambiguity on who "player 1" is. Player 1 is obviously the only controller. This issue has remained invisible to me for a long time because most games seem to happily accept input from any controller. Hades and Bloodstained are doing it differently. They only listen to the first slot.

So I reconfigured MoltenGamepad to create only one virtual pad instead of its default four, and suddenly Hades and Bloodstained could use the virtual pad just fine in Proton 6.3, without any configuration changes, command line switches, or environment variables. What I tested next was MG in its default 4-virtpad configuration, but I used the command socket to move my controller from slot 1 to slot 4. When I did so, again, Hades and Bloodstained worked.

Okay, so controllers are being enumerated in reverse. Is there any way to pinpoint exactly where that is happening and why, and prove that it is explicitly different?

I took advantage of the fact that I have both Flatpak Steam and distro Steam available to me to drill this down. I first noticed gamepads enumerating in reverse when I would check on the controllers in the controller configuration dialog while running Flatpak Steam. MoltenGamepad can forward rumble events, and rumble events are used to identify controllers. Clicking identify on the first slot as identified by Steam, I did not feel a rumble. But on the fourth, I felt a rumble. However, distro Steam does enumerate the controllers in the right order. If I identify the controller in slot 1, I get the rumble as expected.

At this point, I'm going to make this part of this long story short. To get controller input working in Bloodstained:

* In distro Steam, with Proton < 5.13: virtual slot 1
* In distro Steam, with Proton >= 5.13: virtual slot 4
* In Flatpak Steam, with any version of Proton: virtual slot 4

It seems that any time a sandbox is entered, controllers are enumerated backwards. It doesn't seem that sub-sandboxes affect the issue; if they did, then it would go away with Flatpak Steam and anything using pressure-vessel.

Lastly, I just needed a test as simple as possible to prove that current behavior was incorrect, and to prove that it is consistent, and doesn't have anything to do with MoltenGamepad. I used sdl2-jstest for this. In the Flatpak version, virtual slot 1 is enumerated as the last index. In an unsandboxed version, virtual slot 1 is the first index. And then I turned off MoltenGamepad, connected two DualShock 4s, and repeated the test; the first connected controller is the last index in the Flatpak version, and the first index in the unsandboxed version.

Yeah, so that was a whole lot of text to explain that I found that Flatpak, or one of its components, likely bwrap, is doing something wrong that is breaking controller enumeration. Or SDL2 does something wrong in a sandboxed environment. Either way: not your problem.

Problem 2: The udev rules used to hide real devices and show virtual devices did not correctly account for Steam Input.

I found that Guilty Gear Strive was "working" with Steam Input off, for a very loose definition of working. It does read controller inputs just fine, and they do what they seem to be meant to. But it shows keyboard prompts, and only listens to controllers for players 1 and 2. Unlike Battle for the Grid, it does not have a sensible fallback for if you force off Steam Input.

Guilty Gear Strive *has* to have Steam Input. Okay, fine, sure. But I found that anytime I turned it on, in every game, every environment, controllers stopped working.

Well, turns out it's actually my bad, and I feel a little embarrassed about it. Part of what makes MoltenGamepad work in the system-wide configuration is a udev rule that hides real devices from the rest of the system by changing their ownership and permissions, but leaves virtual devices visible. Specifically, *only its own virtual devices.* What I found when I enabled "Xbox Configuration Support" was that the resulting virtual devices were hidden via the udev rules as if MoltenGamepad was going to translate them, but MG already seems to ignore virtual devices. With a manual `chmod a+rw` on the event nodes, Guilty Gear Strive was working, and so were the other games when I had MG configured to make only one virtual controller.

So I needed to fix MoltenGamepad's udev rule. Again, not your problem.

So with this comment, I'll be closing the issue. Device enumeration is certainly not in the scope of the steam runtime specifically; other apps like RPCS3 and GZDoom get messed up by this. So it's more appropriate to make an issue against Flatpak.

As for the broken udev rule, I'll probably have to fork MoltenGamepad and take custodianship of it. The original creator and maintainer is MIA, and the project seems to be abandoned, so making another pull request will probably go nowhere. I'm not comfortable with C++ at all, but I can at least accept pull requests and improve install scripts.

Thanks for your time and patience. :)
